docs/source/runtime_troubles.rst

```diff
@@ -0,0 +1,311 @@
+annotations that would cause runtime errors.
+
+Any type can be entered as a string literal, and you can combine
+string-literal types with non-string-literal types freely:
+
+.. code-block:: python
+
+   def f(a: List['A']) -> None: ...  # OK
+   def g(n: 'int') -> None: ...      # OK, though not useful
+
+   class A: pass
+
+String literal types are never needed in ``# type:`` comments and :ref:`stub files <stub-files>`.
+
+String literal types must be defined (or imported) later *in the same module*.
+They cannot be used to leave cross-module references unresolved.  (For dealing
+with import cycles, see :ref:`import-cycles`.)
+
+.. _future-annotations:
+
+Future annotations import (PEP 563)
+-----------------------------------
+
+Many of the issues described here are caused by Python trying to evaluate
+annotations. From Python 3.10 on, Python will no longer attempt to evaluate
+function and variable annotations. This behaviour is made available in Python
+3.7 and later through the use of ``from __future__ import annotations``.
+
+This can be thought of as automatic string literal-ification of all function and
+variable annotations. Note that function and variable annotations are still
+required to be valid Python syntax. For more details, see :pep:`563`.
+
+.. note::
+
+    Even with the ``__future__`` import, there are some scenarios that could
+    still require string literals or result in errors, typically involving use
+    of forward references or generics in:
+
+    * :ref:`type aliases <type-aliases>`;
+    * :ref:`casts <casts>`;
+    * type definitions (see :py:class:`~typing.TypeVar`, :py:func:`~typing.NewType`, :py:class:`~typing.NamedTuple`);
+    * base classes.
+
+    .. code-block:: python
+
+        # base class example
+        class A(Tuple['B', 'C']): ... # OK
+        class B: ...
+        class C: ...
+
+.. note::
+
+    Some libraries may have use cases for dynamic evaluation of annotations, for
+    instance, through use of ``typing.get_type_hints`` or ``eval``. If your
+    annotation would raise an error when evaluated (say by using :pep:`604`
+    syntax with Python 3.9), you may need to be careful when using such
+    libraries.
+
+.. _typing-type-checking:
+
+typing.TYPE_CHECKING
+--------------------
+
+The :py:mod:`typing` module defines a :py:data:`~typing.TYPE_CHECKING` constant
+that is ``False`` at runtime but treated as ``True`` while type checking.
+
+Since code inside ``if TYPE_CHECKING:`` is not executed at runtime, it provides
+a convenient way to tell mypy something without the code being evaluated at
+runtime. This is most useful for resolving :ref:`import cycles <import-cycles>`.
+
+.. note::
+
+   Python 3.5.1 and below don't have :py:data:`~typing.TYPE_CHECKING`. An
+   alternative is to define a constant named ``MYPY`` that has the value
+   ``False`` at runtime. Mypy considers it to be ``True`` when type checking.
+
+Class name forward references
```

Comment:
This sections seems repetitive with the previous sections

Reply:
Do you have a specific thing we should cut? I do like that this section has somewhat standalone examples; it spells out what fixes should look like for users with this problem.

Reply:
Here's one idea:

* Move this section before the other sections dealing with forward references.
* Add a sentence to the end of the section along the lines of "The followings sections explain future annotations and string literal types in more detail.
* Remove/reduce redundant discussions in the other sections, or add something like "As discussed earlier, ...".